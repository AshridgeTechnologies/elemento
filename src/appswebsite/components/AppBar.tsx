import * as React from 'react';
import Mui_AppBar from '@mui/material/AppBar';
import Box from '@mui/material/Box';
import Toolbar from '@mui/material/Toolbar';
import Typography from '@mui/material/Typography';
import IconButton from '@mui/material/IconButton';
import MenuIcon from '@mui/icons-material/Menu';
import UserMenu from '../../shared/UserMenu'

export default function AppBar() {
    return (
        <Box sx={{ flexGrow: 1 }}>
            <Mui_AppBar position="sticky">
                <Toolbar variant="dense">
                    <IconButton
                        size="large"
                        edge="start"
                        color="inherit"
                        aria-label="menu"
                        sx={{ mr: 2 }}
                    >
                        <MenuIcon />
                    </IconButton>
                    <Typography variant="h6" component="div" sx={{ fontSize: 16, flexGrow: 1 }}>
                        Elemento Apps
                    </Typography>
                    <UserMenu/>
                </Toolbar>
            </Mui_AppBar>
        </Box>
    );
}