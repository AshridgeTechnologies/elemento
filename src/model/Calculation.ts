import {ComponentType, PropertyDef, PropertyExpr, PropertyValueType} from './Types'
import Element from './Element'
import BaseElement, {propDef} from './BaseElement'

type Properties = Partial<Readonly<{
    calculation: PropertyExpr,
    label: PropertyValueType<string>,
    display: PropertyValueType<boolean>,
    width: PropertyValueType<number | string>,
}>>

export default class Calculation extends BaseElement<Properties> implements Element {

    static kind = 'Calculation'
    static get iconClass() { return 'calculate_outlined' }
    type(): ComponentType { return 'statefulUI' }

    get calculation() {return this.properties.calculation}
    get label() {return this.properties.label}
    get display() {return this.properties.display ?? true}
    get width() {return this.properties.width}

    get propertyDefs(): PropertyDef[] {
        return [
            propDef('calculation', 'expr', {state: true}),
            propDef('label', 'string'),
            propDef('display', 'boolean'),
            propDef('width', 'string|number'),
        ]
    }
}