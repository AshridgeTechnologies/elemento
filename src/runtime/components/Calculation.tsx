import React from 'react'
import {SxProps, TextField} from '@mui/material'
import yaml from 'js-yaml'
import {definedPropertiesOf} from '../../util/helpers'
import {PropVal, StylesPropVals, valueOf, valueOfProps} from '../runtimeFunctions'
import {useGetObjectState} from '../appData'
import {BaseComponentState, ComponentState} from './ComponentState'
import {isArray, isObject} from 'radash'
import {equals, pick} from 'ramda'
import {formControlStyles, inputElementProps, propsForInputComponent, sxFieldSetProps, sxProps} from './ComponentHelpers'


type Properties = Readonly<{path: string, label?: PropVal<string>, show?: PropVal<boolean>, styles?: StylesPropVals}>
type ExternalStateProperties = Partial<Readonly<{value: any, whenTrueAction: () => void}>>
type InternalStateProperties = Partial<Readonly<{previousValueTruthy: any}>>

const isObjOrArray = (value: any) => isObject(value) ?? isArray(value)
const formatDisplay = (value: any) => {
    if (typeof value === 'string') return value

    const hasNestedObject = isObjOrArray(value) && Object.values(value).some( isObjOrArray )
    const options = {
        skipInvalid: true,
        flowLevel: hasNestedObject ? 1 : 0
    }

    return yaml.dump(value, options)
}

export default function Calculation({path, ...props}: Properties) {
    const {label, show, styles = {}} = valueOfProps(props)
    const sx = {...sxProps(pick(formControlStyles, styles), show), fieldset: sxFieldSetProps(styles)} as SxProps<{}>

    const state = useGetObjectState<CalculationState>(path)
    const {value} = state
    state.checkTriggered()
    const multiline = true
    const multilineProps = multiline ? {minRows: 1, maxRows: 10} : {}
    const optionalProps = definedPropertiesOf({label, multiline, ...multilineProps})
    const inputComponentProps = propsForInputComponent(undefined, styles)
    const inputProps = inputElementProps(styles, false, {})

    return React.createElement(TextField, {
        id: path,
        type: 'text',
        variant: 'outlined',
        size: 'small',
        value: formatDisplay(value),
        InputLabelProps: {shrink: true},
        sx,
        ...inputProps,
        ...inputComponentProps,
        ...optionalProps
    })
}

export class CalculationState extends BaseComponentState<ExternalStateProperties, InternalStateProperties>
    implements ComponentState<CalculationState> {

    // 'calculation' is called value in the state object to be consistent with input elements
    // AND so it will work with Form
    get value() {
        return valueOf(this.props.value)
    }

    private get previousValueTruthy(): boolean {
        return this.state.previousValueTruthy
    }

    private get whenTrueAction() {
        return this.props.whenTrueAction
    }

    protected isEqualTo(newObj: this): boolean {
        return equals(this.props.value, newObj.props.value) && this.props.whenTrueAction === newObj.props.whenTrueAction
    }

    checkTriggered() {
        if (!this.whenTrueAction) return
        const currentValueTruthy = !!this.value
        const triggered = currentValueTruthy && !this.previousValueTruthy
        if (triggered) {
            this.whenTrueAction()
        }
        if (this.previousValueTruthy !== currentValueTruthy) {
            this.updateState({previousValueTruthy: currentValueTruthy})
        }
    }

    valueOf() {
        return this.value
    }

    toString() {
        return this.value.toString()
    }
}

Calculation.State = CalculationState
