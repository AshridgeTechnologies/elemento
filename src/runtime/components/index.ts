export {default as App} from './App'
export {default as Button} from './Button'
export {default as Icon} from './Icon'
export {default as Image} from './Image'
export {default as UserLogon} from './UserLogon'
export {default as Menu} from './Menu'
export {default as MenuItem} from './MenuItem'
export {default as NumberInput} from './NumberInput'
export {default as DateInput} from './DateInput'
export {default as TrueFalseInput} from './TrueFalseInput'
export {default as SpeechInput} from './SpeechInput'
export {default as Data} from './Data'
export {default as Calculation} from './Calculation'
export {default as Timer} from './Timer'
export {default as Collection} from './Collection'
export {default as Page} from './Page'
export {default as ComponentElement} from './ComponentElement'
export {default as Form} from './Form'
export {default as BaseFormState} from './FormState'
export {default as SelectInput} from './SelectInput'
export {default as TextElement} from './TextElement'
export {default as Layout} from './Layout'
export {default as Block} from './Block'
export {default as AppBar} from './AppBar'
export {default as ListElement} from './ListElement'
export {default as ListItem} from './ListItem'
export {default as ItemSet} from './ItemSet'
export {default as ItemSetItem} from './ItemSetItem'
export {default as TextInput} from './TextInput'
export {default as MemoryDataStore} from './MemoryDataStore'
export {default as FileDataStore} from './FileDataStore'
export {default as FileDataStoreImpl} from './FileDataStoreImpl'
export {default as IdbDataStoreImpl} from './IdbDataStoreImpl'
export {default as BrowserDataStore} from './BrowserDataStore'
export {default as FirestoreDataStore} from './FirestoreDataStore'
export {default as FirestoreDataStoreImpl} from './FirestoreDataStoreImpl'
export {default as ServerAppConnector} from './ServerAppConnector'
export {default as Adapter} from './Adapter'
