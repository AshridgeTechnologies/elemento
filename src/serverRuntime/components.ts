export {default as FirestoreDataStore} from './FirestoreDataStore'
export {default as Collection} from './Collection'

// NOTE: Must add the name of each component to ./names.ts