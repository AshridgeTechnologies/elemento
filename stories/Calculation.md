Calculation
===========

Aims
----

- Reusable calculations on page
- Split up calculations
- Display readonly data
- Display data temporarily for debugging/tutorials

Requirements
------------

- Calculation model element
- Formula property - Formula only
- Display property, default true, hides if false
- Width property
- State object with Value
- Displays like readonly text input
- Expands to as many lines as needed to show data - min 1

Technical
---------

- Similar to Data, without Set