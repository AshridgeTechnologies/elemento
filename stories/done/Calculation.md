Calculation
===========

Aims
----

- Reusable calculations on page
- Split up calculations
- Display readonly data
- Display data temporarily for debugging/tutorials

Requirements
------------

- ✅ Calculation model element
- ✅ Calculation property - Formula only
- ✅ Display property, default true, hides if false
- ✅ Width property
- ✅ Label property
- ✅ State object with Value
- ✅ Displays like readonly text input
- ✅ Expands to as many lines as needed to show data - min 1
- ✅ Values displayed in readable form

Technical
---------

- Similar to Data, without Set