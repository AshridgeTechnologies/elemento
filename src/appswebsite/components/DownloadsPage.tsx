import React from 'react'
import {Box, Container, Link, Stack, Typography} from '@mui/material'
import AppBar from '../../shared/AppBar'
import {theme} from '../../shared/styling'
import {ThemeProvider} from '@mui/material/styles'

export default function LearnPage() {

    return <ThemeProvider theme={theme}>
        <Box display='flex' flexDirection='column' height='100%' width='100%'>
            <Box flex='0'>
                <AppBar title='Elemento Online - Downloads' userMenu={false}/>
            </Box>
            <Box flex='1' minHeight={0}>
                <Container>
                    <Stack maxWidth={600} spacing={3} mt={3}>
                        <Typography variant='h4'>Elemento Development Server</Typography>
                        <Typography>Please download the correct version for your operating system.</Typography>
                        <Typography>Your computer will complain that the app is not known, approved, signed etc.<br/>
                            We do not currently have the resources to jump through all the hoops required to get the app
                            approved on each operating system,
                            so you will just have to trust us here, and continue past all the warnings.
                        </Typography>
                        <Link href='/devServer/win/ElementoDevServer.exe'><Typography variant='h5'>Windows</Typography></Link>
                        <Link href='/devServer/macos/ElementoDevServer.zip'><Typography variant='h5'>Mac</Typography></Link>
                        <Link href='/devServer/linux/ElementoDevServer'><Typography variant='h5'>Linux</Typography></Link>
                    </Stack>
                </Container>
            </Box>
        </Box>
    </ThemeProvider>
}
