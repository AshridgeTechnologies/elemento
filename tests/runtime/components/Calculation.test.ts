/**
 * @jest-environment jsdom
 */

import {Calculation} from '../../../src/runtime/components/index'
import {testAppInterface, valueObj, wrappedTestElement} from '../../testutil/testHelpers'
import {render} from '@testing-library/react'
import '@testing-library/jest-dom'
import {testContainer} from '../../testutil/rtlHelpers'
import {CalculationState} from '../../../src/runtime/components/Calculation'

const [calculation] = wrappedTestElement(Calculation, CalculationState)

test('Calculation element produces output with properties supplied', () => {
    const {container} = render(calculation('app.page1.width', {value: 'Hi there!'}, {
            styles: {width: 23},
            label: 'Item Description'
        }))
    expect(container.innerHTML).toMatchSnapshot()
})

test('Calculation element produces output with multiline', () => {
    const {container} = render(calculation('app.page1.description', {value: 'Hi there!\nHow are you?'}, {label: 'Item Description'}))
    expect(container.innerHTML).toMatchSnapshot()
})

test('Calculation element produces output with default values where properties omitted',
    () => {
        const {container} = render(calculation('app.page1.width', {}))
        expect(container.innerHTML).toMatchSnapshot()
    }
)

test('Calculation shows value from the state supplied', () => {
    const {el} = testContainer(calculation('app.page1.widget1', {value:'Hello!'}, {}))
    expect(el`app.page1.widget1`.textContent).toBe('Hello!')
})

test('Calculation shows empty value when state value is absent', () => {
    const {el} = testContainer(calculation('app.page1.widget1', {}))
    expect(el`app.page1.widget1`.textContent).toBe('')
})

test('State class has correct properties and functions', () => {
    const state = new Calculation.State({value: {a: 10, b: 'bee'}})
    const appInterface = testAppInterface(); state.init(appInterface, 'testPath')
    expect(state.value).toStrictEqual({a: 10, b: 'bee'})
})

test('State class gives correct value when its value is a value object', () => {
    const state = new Calculation.State({value: valueObj('car')})
    expect(state.value).toBe('car')
})

test('string conversion uses the value toString', () => {
    const value = [1, 2, 3]
    const state = new Calculation.State({value})
    expect('x' + state).toBe('x1,2,3')
})

