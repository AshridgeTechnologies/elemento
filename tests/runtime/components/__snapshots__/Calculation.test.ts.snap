// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Calculation element produces output with default values where properties omitted 1`] = `"<div class="MuiFormControl-root MuiTextField-root css-55ifni-MuiFormControl-root-MuiTextField-root"><div class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall MuiInputBase-multiline css-171086r-MuiInputBase-root-MuiOutlinedInput-root"><textarea aria-invalid="false" id="app.page1.width" class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline MuiInputBase-inputSizeSmall css-msbelq-MuiInputBase-input-MuiOutlinedInput-input" style="height: 0px; overflow: hidden;"></textarea><textarea aria-hidden="true" class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline MuiInputBase-inputSizeSmall css-msbelq-MuiInputBase-input-MuiOutlinedInput-input" readonly="" tabindex="-1" style="visibility: hidden; position: absolute; overflow: hidden; height: 0px; top: 0px; left: 0px; transform: translateZ(0); padding-top: 0px; padding-bottom: 0px; width: 100%;"></textarea><fieldset aria-hidden="true" class="MuiOutlinedInput-notchedOutline css-18p5xg2-MuiNotchedOutlined-root-MuiOutlinedInput-notchedOutline"><legend class="css-1nf2c5d-MuiNotchedOutlined-root"><span class="notranslate" aria-hidden="true">â€‹</span></legend></fieldset></div></div>"`;

exports[`Calculation element produces output with multiline 1`] = `
"<div class="MuiFormControl-root MuiTextField-root css-55ifni-MuiFormControl-root-MuiTextField-root"><label class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-outlined css-hvjq6j-MuiFormLabel-root-MuiInputLabel-root" data-shrink="true" for="app.page1.description" id="app.page1.description-label">Item Description</label><div class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall MuiInputBase-multiline css-171086r-MuiInputBase-root-MuiOutlinedInput-root"><textarea aria-invalid="false" id="app.page1.description" class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline MuiInputBase-inputSizeSmall css-msbelq-MuiInputBase-input-MuiOutlinedInput-input" style="height: 0px; overflow: hidden;">Hi there!
How are you?</textarea><textarea aria-hidden="true" class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline MuiInputBase-inputSizeSmall css-msbelq-MuiInputBase-input-MuiOutlinedInput-input" readonly="" tabindex="-1" style="visibility: hidden; position: absolute; overflow: hidden; height: 0px; top: 0px; left: 0px; transform: translateZ(0); padding-top: 0px; padding-bottom: 0px; width: 100%;"></textarea><fieldset aria-hidden="true" class="MuiOutlinedInput-notchedOutline css-18p5xg2-MuiNotchedOutlined-root-MuiOutlinedInput-notchedOutline"><legend class="css-ex8a5f-MuiNotchedOutlined-root"><span>Item Description</span></legend></fieldset></div></div>"
`;

exports[`Calculation element produces output with properties supplied 1`] = `"<div class="MuiFormControl-root MuiTextField-root css-aty329-MuiFormControl-root-MuiTextField-root"><label class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-outlined css-hvjq6j-MuiFormLabel-root-MuiInputLabel-root" data-shrink="true" for="app.page1.width" id="app.page1.width-label">Item Description</label><div class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall MuiInputBase-multiline css-171086r-MuiInputBase-root-MuiOutlinedInput-root"><textarea aria-invalid="false" id="app.page1.width" class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline MuiInputBase-inputSizeSmall css-msbelq-MuiInputBase-input-MuiOutlinedInput-input" style="height: 0px; overflow: hidden;">Hi there!</textarea><textarea aria-hidden="true" class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline MuiInputBase-inputSizeSmall css-msbelq-MuiInputBase-input-MuiOutlinedInput-input" readonly="" tabindex="-1" style="visibility: hidden; position: absolute; overflow: hidden; height: 0px; top: 0px; left: 0px; transform: translateZ(0); padding-top: 0px; padding-bottom: 0px; width: 100%;"></textarea><fieldset aria-hidden="true" class="MuiOutlinedInput-notchedOutline css-18p5xg2-MuiNotchedOutlined-root-MuiOutlinedInput-notchedOutline"><legend class="css-ex8a5f-MuiNotchedOutlined-root"><span>Item Description</span></legend></fieldset></div></div>"`;
